File: fsharp/FsFind/FsFind/FindOptions.fs
﻿namespace FsFind

open System
open System.Text.Json
open System.Text.RegularExpressions
open System.Xml.Linq

module FindOptions =   

    type FindOption = {
        ShortArg : string;
        LongArg : string;
        Description : string
    }

    let AddExtensions (exts : string) (extList : string list) : string list =
        List.append extList (FileUtil.ExtensionsListFromString exts)

    let AddPattern (pattern : string) (patternList : Regex list) : Regex list =
        List.append patternList [Regex(pattern)]

    let FileTypesListFromString (fts : string) : FileType list =
        let nonWord = Regex(@"\W+")
        nonWord.Split(fts)
        |> Array.toList
        |> List.filter (fun (x : string) -> String.IsNullOrEmpty(x) = false)
        |> List.map (fun (x : string) -> FileTypes.FromName x)

    let AddFileTypes (fts : string) (ftList : FileType list) : FileType list =
        List.append ftList (FileTypesListFromString fts)

    let argActionMap : Map<string, string -> FindSettings.t -> FindSettings.t> =
        [
            ("encoding", (fun (s : string) (settings : FindSettings.t) -> { settings with TextFileEncoding = s }));
            ("in-archiveext", (fun (s : string) (settings : FindSettings.t) -> { settings with InArchiveExtensions = AddExtensions s settings.InArchiveExtensions }));
            ("in-archivefilepattern", (fun (s : string) (settings : FindSettings.t) -> { settings with InArchiveFilePatterns = AddPattern s settings.InArchiveFilePatterns }));
            ("in-dirpattern", (fun (s : string) (settings : FindSettings.t) -> { settings with InDirPatterns = AddPattern s settings.InDirPatterns }));
            ("in-ext", (fun (s : string) (settings : FindSettings.t) -> { settings with InExtensions = AddExtensions s settings.InExtensions }));
            ("in-filepattern", (fun (s : string) (settings : FindSettings.t) -> { settings with InFilePatterns = AddPattern s settings.InFilePatterns }));
            ("in-filetype", (fun (s : string) (settings : FindSettings.t) -> { settings with InFileTypes = AddFileTypes s settings.InFileTypes }));
            ("in-linesafterpattern", (fun (s : string) (settings : FindSettings.t) -> { settings with InLinesAfterPatterns = AddPattern s settings.InLinesAfterPatterns }));
            ("in-linesbeforepattern", (fun (s : string) (settings : FindSettings.t) -> { settings with InLinesBeforePatterns = AddPattern s settings.InLinesBeforePatterns }));
            ("linesafter", (fun (s : string) (settings : FindSettings.t) -> { settings with LinesAfter = Int32.Parse(s) }));
            ("linesaftertopattern", (fun (s : string) (settings : FindSettings.t) -> { settings with LinesAfterToPatterns = AddPattern s settings.LinesAfterToPatterns }));
            ("linesafteruntilpattern", (fun (s : string) (settings : FindSettings.t) -> { settings with LinesAfterUntilPatterns = AddPattern s settings.LinesAfterUntilPatterns }));
            ("linesbefore", (fun (s : string) (settings : FindSettings.t) -> { settings with LinesBefore = Int32.Parse(s) }));
            ("maxlinelength", (fun (s : string) (settings : FindSettings.t) -> { settings with MaxLineLength = Int32.Parse(s) }));
            ("out-archiveext", (fun (s : string) (settings : FindSettings.t) -> { settings with OutArchiveExtensions = AddExtensions s settings.OutArchiveExtensions }));
            ("out-archivefilepattern", (fun (s : string) (settings : FindSettings.t) -> { settings with OutArchiveFilePatterns = AddPattern s settings.OutArchiveFilePatterns }));
            ("out-dirpattern", (fun (s : string) (settings : FindSettings.t) -> { settings with OutDirPatterns = AddPattern s settings.OutDirPatterns }));
            ("out-ext", (fun (s : string) (settings : FindSettings.t) -> { settings with OutExtensions = AddExtensions s settings.OutExtensions }));
            ("out-filepattern", (fun (s : string) (settings : FindSettings.t) -> { settings with OutFilePatterns = AddPattern s settings.OutFilePatterns }));
            ("out-filetype", (fun (s : string) (settings : FindSettings.t) -> { settings with OutFileTypes = AddFileTypes s settings.OutFileTypes }));
            ("out-linesafterpattern", (fun (s : string) (settings : FindSettings.t) -> { settings with OutLinesAfterPatterns = AddPattern s settings.OutLinesAfterPatterns }));
            ("out-linesbeforepattern", (fun (s : string) (settings : FindSettings.t) -> { settings with OutLinesBeforePatterns = AddPattern s settings.OutLinesBeforePatterns }));
            ("findpattern", (fun (s : string) (settings : FindSettings.t) -> { settings with FindPatterns = AddPattern s settings.FindPatterns }));
        ] |> Map.ofList

    let flagActionMap : Map<string, bool -> FindSettings.t -> FindSettings.t> =
        [
            ("allmatches", (fun (b : bool) (settings : FindSettings.t) -> { settings with FirstMatch = not b }));
            ("archivesonly", (fun (b : bool) (settings : FindSettings.t) -> FindSettings.SetArchivesOnly b settings));
            ("colorize", (fun (b : bool) (settings : FindSettings.t) -> { settings with Colorize = b }));
            ("debug", (fun (b : bool) (settings : FindSettings.t) -> FindSettings.SetDebug b settings));
            ("excludehidden", (fun (b : bool) (settings : FindSettings.t) -> { settings with ExcludeHidden = b }));
            ("firstmatch", (fun (b : bool) (settings : FindSettings.t) -> { settings with FirstMatch = b }));
            ("help", (fun (b : bool) (settings : FindSettings.t) -> { settings with PrintUsage = b }));
            ("includehidden", (fun (b : bool) (settings : FindSettings.t) -> { settings with ExcludeHidden = not b }));
            ("listdirs", (fun (b : bool) (settings : FindSettings.t) -> { settings with ListDirs = b }));
            ("listfiles", (fun (b : bool) (settings : FindSettings.t) -> { settings with ListFiles = b }));
            ("listlines", (fun (b : bool) (settings : FindSettings.t) -> { settings with ListLines = b }));
            ("multilineoption-REMOVE", (fun (b : bool) (settings : FindSettings.t) -> { settings with MultiLineFind = b }));
            ("nocolorize", (fun (b : bool) (settings : FindSettings.t) -> { settings with Colorize = not b }));
            ("noprintmatches", (fun (b : bool) (settings : FindSettings.t) -> { settings with PrintResults = not b }));
            ("norecursive", (fun (b : bool) (settings : FindSettings.t) -> { settings with Recursive = not b }));
            ("nofindarchives", (fun (b : bool) (settings : FindSettings.t) -> { settings with FindArchives = not b }));
            ("printmatches", (fun (b : bool) (settings : FindSettings.t) -> { settings with PrintResults = b }));
            ("recursive", (fun (b : bool) (settings : FindSettings.t) -> { settings with Recursive = b }));
            ("findarchives", (fun (b : bool) (settings : FindSettings.t) -> { settings with FindArchives = b }));
            ("uniquelines", (fun (b : bool) (settings : FindSettings.t) -> { settings with UniqueLines = b }));
            ("verbose", (fun (b : bool) (settings : FindSettings.t) -> { settings with Verbose = b }));
            ("version", (fun (b : bool) (settings : FindSettings.t) -> { settings with PrintVersion = b }));
        ] |> Map.ofList;

    type FindOptionsDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<System.Collections.Generic.Dictionary<string,string>>>

    let OptionsFromJson (jsonString : string) : FindOption list =
        let findOptionsDict = JsonSerializer.Deserialize<FindOptionsDictionary>(jsonString)
        let optionDicts = findOptionsDict.["findoptions"]
        [ for optionDict in optionDicts do
            let longArg = optionDict.["long"]
            let shortArg = if optionDict.ContainsKey("short") then optionDict.["short"] else ""
            let desc = optionDict.["desc"]
            yield { ShortArg=shortArg; LongArg=longArg; Description=desc } ]

    let OptionsFromXml (xmlString : string) : FindOption list =
        let rec recOptionsFromXml (nodeList : XElement list) (options : FindOption list) : FindOption list =
            match nodeList with
            | [] -> options
            | n :: ns ->
                let short = [for a in n.Attributes(XName.Get("short")) do yield a.Value].Head
                let long = [for a in n.Attributes(XName.Get("long")) do yield a.Value].Head
                let desc = n.Value.Trim()
                recOptionsFromXml ns (List.append options [{ ShortArg=short; LongArg=long; Description=desc }])
        let optNodes = XDocument.Parse(xmlString).Descendants(XName.Get("findoption"))
        recOptionsFromXml (List.ofSeq optNodes) []
        |> List.sortBy (fun o -> if (o.ShortArg <> "") then (o.ShortArg.ToLower() + "@" + o.LongArg) else o.LongArg)

//    let _findOptionsResource = EmbeddedResource.GetResourceFileContents("FsFind.Resources.findoptions.xml");
    let _findOptionsResource = EmbeddedResource.GetResourceFileContents("FsFind.Resources.findoptions.json");
//    let options = OptionsFromXml(_findOptionsResource)
    let options = OptionsFromJson(_findOptionsResource)

    let SettingsFromArgs (args : string[]) : FindSettings.t * string =
        let optionNameMap =
            let shortargs = seq { for opt in options do if opt.ShortArg <> "" then yield (opt.ShortArg, opt.LongArg) }
            let longargs =  seq { for opt in options do yield (opt.LongArg, opt.LongArg) }
            Seq.append shortargs longargs
            |> Map.ofSeq

        let argRegex = Regex("^(?:-{1,2})(?<opt>.*)$")
        
        let (|IsOption|_|) (arg:string) =            
            let m = argRegex.Match(arg)
            if m.Success then Some(m.Groups.["opt"].Value) else None

        let rec recSettingsFromArgs (argList : string list) (settings : FindSettings.t) : FindSettings.t * string =
            match argList with
            | [] -> settings, ""
            | head :: tail ->
                match head with
                | IsOption opt ->
                    if optionNameMap.ContainsKey(opt) then
                        let long = optionNameMap.[opt]
                        if argActionMap.ContainsKey(long) then
                            match tail with
                            | [] ->
                                settings, sprintf "Missing value for option: %s" opt
                            | aHead :: aTail -> 
                                recSettingsFromArgs aTail (argActionMap.[long] aHead settings)
                        elif flagActionMap.ContainsKey(long) then
                            if long = "help" then
                                recSettingsFromArgs [] (flagActionMap.[long] true settings)
                            else
                                recSettingsFromArgs tail (flagActionMap.[long] true settings)
                        else
                            settings, sprintf "Invalid option: %s" opt
                    else
                        settings, sprintf "Invalid option: %s" opt
                | _ -> recSettingsFromArgs tail { settings with StartPath = head }
        recSettingsFromArgs (Array.toList args) { FindSettings.DefaultSettings with PrintResults = true }

    let GetUsageString () : string =
        let optStringMap =
            [ for opt in options do
                let shortstring : string = 
                    if opt.ShortArg <> "" then "-" + opt.ShortArg + ","
                    else ""
                let longstring : string = "--" + opt.LongArg
                yield (opt.LongArg, shortstring + longstring) ]
            |> Map.ofList

        let optDescMap =
            [ for opt in options do
                yield (opt.LongArg, opt.Description) ]
            |> Map.ofList

        let optStrings : string list =
            optStringMap
            |> Map.toList
            |> List.map snd

        let longest = 
            optStrings
            |> Seq.map (fun o -> o.Length)
            |> Seq.max

        let format = " {0,-" + longest.ToString() + "}  {1}"

        let usageStrings =
            [for o in options do
                yield String.Format(format, optStringMap.[o.LongArg], optDescMap.[o.LongArg])]

        let usageString = 
            usageStrings
            |> List.append ["\nUsage:"; " fsfind [options] -s <findpattern> <startpath>\n"; "Options:"] 
            |> String.concat "\n"
        usageString

    let Usage (exitCode : int) : unit =
        let usageString = GetUsageString()
        printfn "%s\n" usageString
        Environment.Exit(exitCode)

File: ocaml/mlfind/src/findoptions.ml
open Core.Std
open Xml;;
(* 
    NOTE: xml parsing is done using 3rd-party library xml-light, which must be installed:

    $ opam install xml-light

    and a package reference must be added to the build line:

    -package xml-light
 *)
type findOption = { long : string; short : string; desc : string };;

let get_desc option_node = 
  match (List.hd (Xml.children option_node)) with
  | Some hd -> String.strip (Xml.pcdata hd)
  | None    -> "";;

let get_findoptions : findOption list = 
  let x = Xml.parse_file (Config.xfindpath ^ "/shared/findoptions.xml") in
  let option_nodes = Xml.children x in
  List.map option_nodes
    ~f:(fun o -> { short=(Xml.attrib o "short"); long=(Xml.attrib o "long"); desc=(get_desc o) });;

let cmp_findoptions so1 so2 = 
  let short_or_long so =
    match (so.short, so.long) with
    | ("", l) -> l
    | (s, l)  -> (String.lowercase s) ^ "@" ^ l in
  let sl1 = short_or_long so1 in
  let sl2 = short_or_long so2 in
  if sl1 < sl2 then -1
  else if sl1 > sl2 then 1
  else 0

let get_opt_strings findoptions = 
  let short_and_or_long so =
    match (so.short, so.long) with
    | ("", l) -> "--" ^ l
    | (s, l)  -> "-" ^ s ^ ",--" ^ l in
  List.map findoptions ~f:short_and_or_long

let pad_string s len =
  if String.length s < len then s ^ (String.make (len - (String.length s)) ' ')
  else s

let rec zip l1 l2 zipped = 
  match (l1, l2) with
  | ([], _) -> zipped
  | (_, []) -> zipped
  | ((h1 :: t1), (h2 :: t2)) -> zip t1 t2 (List.append zipped [(h1, h2)])

(* findoptions_to_string :: [FindOption] -> String *)
let findoptions_to_string findoptions = 
  let sorted = List.sort findoptions ~cmp:cmp_findoptions in
  let opt_strings = get_opt_strings sorted in
  let opt_descs = List.map sorted ~f:(fun so -> so.desc) in
  let longest = List.fold ~init:0 ~f:max (List.map opt_strings ~f:String.length) in
  let format_opt_line o d =
    " " ^ (pad_string o longest) ^ "  " ^ d in
  let zipped = zip opt_strings opt_descs [] in
  let lines = List.map zipped ~f:(fun (o, d) -> format_opt_line o d) in
  String.concat lines ~sep:"\n"

let get_usage findoptions = 
  "Usage:\n mlfind [options] -s <findpattern> <startpath>\n\nOptions:\n"
  ^ (findoptions_to_string findoptions)


type argAction = string -> Findsettings.t -> Findsettings.t;;

let arg_actions : (string * argAction) list = [
  ("in-archiveext", fun s ss -> { ss with in_archiveextensions=Findsettings.add_extensions s ss.in_archiveextensions });
  ("in-archivefilepattern", fun s ss -> { ss with in_archivefilepatterns=List.append ss.in_archivefilepatterns [Re2.Regex.create_exn s] });
  ("in-dirpattern", fun s ss -> { ss with in_dirpatterns=List.append ss.in_dirpatterns [Re2.Regex.create_exn s] });
  ("in-ext", fun s ss -> { ss with in_extensions=Findsettings.add_extensions s ss.in_extensions });
  ("in-filepattern", fun s ss -> { ss with in_filepatterns=List.append ss.in_filepatterns [Re2.Regex.create_exn s] });
  ("out-archiveext", fun s ss -> { ss with out_archiveextensions=Findsettings.add_extensions s ss.out_archiveextensions });
  ("out-archivefilepattern", fun s ss -> { ss with out_archivefilepatterns=List.append ss.out_archivefilepatterns [Re2.Regex.create_exn s] });
  ("out-dirpattern", fun s ss -> { ss with out_dirpatterns=List.append ss.out_dirpatterns [Re2.Regex.create_exn s] });
  ("out-ext", fun s ss -> { ss with out_extensions=Findsettings.add_extensions s ss.out_extensions });
  ("out-filepattern", fun s ss -> { ss with out_filepatterns=List.append ss.out_filepatterns [Re2.Regex.create_exn s] });
  ("path", fun s ss -> { ss with paths=List.append ss.paths s })
];;

type boolFlagAction = bool -> Findsettings.t -> Findsettings.t;;

let bool_flag_actions : (string * boolFlagAction) list = [
  ("archivesonly", fun b ss -> FindSettings.set_archivesonly ss b);
  ("colorize", fun b ss -> { ss with colorize=b });
  ("debug", fun b ss -> FindSettings.set_debug ss b);
  ("excludearchives", fun b ss -> { ss with includearchives=(not b) });
  ("excludehidden", fun b ss -> { ss with excludehidden=b });
  ("help", fun b ss -> { ss with printusage=b });
  ("includearchives", fun b ss -> { ss with includearchives=b });
  ("includehidden", fun b ss -> { ss with excludehidden=(not b) });
  ("listdirs", fun b ss -> { ss with listdirs=b });
  ("listfiles", fun b ss -> { ss with listfiles=b });
  ("nocolorize", fun b ss -> { ss with colorize=(not b) });
  ("noprintmatches", fun b ss -> { ss with printresults=(not b) });
  ("norecursive", fun b ss -> { ss with recursive=(not b) });
  ("printmatches", fun b ss -> { ss with printresults=b });
  ("recursive", fun b ss -> { ss with recursive=b });
  ("verbose", fun b ss -> { ss with verbose=b });
  ("version", fun b ss -> { ss with printversion=b })
];;

let rec arg_name arg = 
  if arg.[0] = '-' && (String.length arg) > 1 then arg_name (String.sub arg ~pos:1 ~len:((String.length arg) - 1))
  else arg

let get_long_arg findoptions arg = 
  match List.find findoptions ~f:(fun o -> o.long = arg || o.short = arg) with
  | Some opt -> Some opt.long
  | None     -> None

let settings_from_args findoptions args = 
  let rec rec_settings_from_args (settings : Findsettings.t) (args : string list) =
    match args with
    | [] -> Ok settings
    | hd :: tl when hd.[0] = '-' ->
        (let arg = arg_name hd in
         match get_long_arg findoptions arg with
         | Some "help" -> rec_settings_from_args { settings with printusage=true } []
         | Some long_arg ->
             (match List.find bool_flag_actions ~f:(fun (s, _) -> s = long_arg) with
              | Some (_, f) -> rec_settings_from_args (f true settings) tl
              | None ->
                (match List.find arg_actions ~f:(fun (s, _) -> s = long_arg) with
                 | Some (_, f) ->
                    (match tl with
                     | [] -> Error (sprintf "Missing value for option: %s" arg)
                     | h :: t  -> rec_settings_from_args (f h settings) t)
                 | None -> Error (sprintf "Invalid option: %s" arg)))
         | None -> Error (sprintf "Invalid option: %s" arg))
    | hd :: tl -> rec_settings_from_args { settings with paths=List.append settings.paths hd } tl in
  (* default listfiles to true since running as cli *)
  rec_settings_from_args { Findsettings.default_settings with listfiles=true } args;;

